pymc/sampling.py

```diff
@@ -1534,6 +1542,166 @@ def stop_tuning(step):
+                    )  # Mixture distributions have RandomVariable parameters that are not measurable (not in model.basic_RVs). These shouldn't propagate their volatility
+                )
+            )
+        ]
+
+    # We need a function graph to walk the clients and propagate the volatile property
+    fg = FunctionGraph(outputs=outputs, clone=False)
+
+    # Walk the graph from inputs to outputs and tag the volatile variables
+    nodes: List[Variable] = general_toposort(
+        fg.outputs, deps=lambda x: x.owner.inputs if x.owner else []
+    )
+    volatile_nodes: Set[Any] = set()
+    for node in nodes:
+        if isinstance(node, SharedVariable) or node in fg.outputs or node in givens_dict:
+            volatile_nodes.add(node)
+        is_volatile = node in volatile_nodes
+
+        # Propagate the volatile property to the node's clients
+        if is_volatile:
+            clients = [out for c, _ in fg.clients[node] if c != "output" for out in c.outputs]  # type: ignore
+            for client in clients:
+                if is_a_random_variable(client) and any(
+                    rv_param in volatile_nodes
+                    for rv_param in distribution_not_dist_parameters(client)
+                ):
+                    # Random variables that have volatile parameters are also volatile
+                    volatile_nodes.add(client)
+                elif not (client in vars_in_trace and is_a_random_variable(client)):
+                    # Other variables that aren't random variables in the trace are volatile
+                    volatile_nodes.add(client)
+
+    # Collect the function inputs by walking the graph from the outputs. Inputs will be:
+    # 1. Random variables that are not volatile
+    # 2. Variables that have no owner and are not constant or shared
+    inputs = []
+
+    def expand(node):
+        if (
+            (
+                node.owner is None and not isinstance(node, (Constant, SharedVariable))
+            )  # Variables without owners that are not constant or shared
+            or node in vars_in_trace  # Variables in the trace
+        ) and node not in volatile_nodes:
+            # This test will include variables without owners, and that are not constant
+            # or shared, because these nodes will never be considered volatile
+            inputs.append(node)
+        if node.owner:
+            return node.owner.inputs
+
+    # walk produces a generator, so we have to actually exhaust the generator in a list to walk
+    # the entire graph
+    list(walk(fg.outputs, expand))
+
+    # Populate the givens list
+    givens = [
+        (
+            node,
+            value
+            if isinstance(value, (Variable, Apply))
+            else at.constant(value, dtype=getattr(node, "dtype", None), name=node.name),
+        )
+        for node, value in givens_dict.items()
+    ]
+
+    return compile_pymc(inputs, fg.outputs, givens=givens, on_unused_input="ignore", **kwargs)
```

Comment:
@lucianopaz what about this? It's very similar but it is more strict about the sources of `volatility`, and after that any node that has a volatile input is considered volatile, regardless of what it is.

The reason I went to look for an alternative is the case where `dist`s depend on other standard RVs. With this code I think the following model would work as expected:

```python
with pm.Model() as m:
  latent_mu = pm.Normal("latent_mu")
  base_dist = pm.Normal.dist(latent_mu, 1)
  llike = pm.Censored("llike", base_dist, lower=-1, upper=1, observed=data)
```

```suggestion
    if givens_dict is None:
        givens_dict = {}

    if basic_rvs is None:
        basic_rvs = []

    # We need a function graph to walk the clients and propagate the volatile property
    fg = FunctionGraph(outputs=outputs, clone=False)

    # Walk the graph from inputs to outputs and tag the volatile variables
    nodes: List[Variable] = general_toposort(
        fg.outputs, deps=lambda x: x.owner.inputs if x.owner else []
    )
    volatile_nodes: Set[Any] = set()
    for node in nodes:
        if (
            # SharedVariables, except RandomState/Generators
            (isinstance(node, SharedVariable) and not isinstance(node, (RandomStateSharedVariable, RandomGeneratorSharedVariable)))
            # Basic RVs that are not in the trace
            or (node.owner and isinstance(node.owner.op, RandomVariable) and node not in vars_in_trace and node in basic_rvs)
            # Variables that have any volatile input
            or (node.owner and any(inp in volatile_nodes for inp in node.owner.inputs))
            or node in fg.outputs
            or node in givens_dict
        ):
            volatile_nodes.add(node)

    # Collect the function inputs by walking the graph from the outputs. Inputs will be:
    # 1. Random variables that are not volatile
    # 2. Variables that have no owner and are not constant or shared
    inputs = []
    def expand(node):
        if (
            (
                node.owner is None and not isinstance(node, (Constant, SharedVariable))
            )  # Variables without owners that are not constant or shared
            or node in vars_in_trace  # Variables in the trace
        ) and node not in volatile_nodes:
            # This test will include variables without owners, and that are not constant
            # or shared, because these nodes will never be considered volatile
            inputs.append(node)
        if node.owner:
            return node.owner.inputs

    # walk produces a generator, so we have to actually exhaust the generator in a list to walk
    # the entire graph
    list(walk(fg.outputs, expand))

    # Populate the givens list
    givens = [
        (
            node,
            value
            if isinstance(value, (Variable, Apply))
            else at.constant(value, dtype=getattr(node, "dtype", None), name=node.name),
        )
        for node, value in givens_dict.items()
    ]

    return compile_pymc(inputs, fg.outputs, givens=givens, on_unused_input="ignore", **kwargs)
```